### PR TITLE
[MIG][11.0] project_wbs

### PR DESCRIPTION
**Pull request by AaronHForgeFlow**
Just one change respect to previous version. I included the filters in the analytic account for the different account classes, this is, projects, deliverables and work packages.

IMO this should be included also in v10 and I'll create a Pull Request for that. At the beginning there were specific menus for every account class. In v10 we decided not to add those menus. However, currently in v10 it is hard to filter by account class. Adding filters should be enough.

### CODE REVIEW

**Comment by JordiBForgeFlow** [comment]
@aheficent there are some warnings in Runbot:
`2018-08-13 13:15:43,460 185 WARNING openerp_test odoo.addons.base.ir.ir_ui_view: Error-prone use of @class in view project.project.kanban.wbs (): use the hasclass(*classes) function to filter elements by their classes
2018-08-13 13:15:43,460 185 WARNING openerp_test odoo.addons.base.ir.ir_ui_view: Error-prone use of @class in view project.project.kanban.wbs (): use the hasclass(*classes) function to filter elements by their classes
2018-08-13 13:15:43,923 185 WARNING openerp_test odoo.addons.base.ir.ir_ui_view: Error-prone use of @class in view project.project.kanban.wbs (project_wbs.view_project_kanban): use the hasclass(*classes) function to filter elements by their classes
2018-08-13 13:15:43,923 185 WARNING openerp_test odoo.addons.base.ir.ir_ui_view: Error-prone use of @class in view project.project.kanban.wbs (project_wbs.view_project_kanban): use the hasclass(*classes) function to filter elements by their classes`

**Comment by AaronHForgeFlow** [comment]
Last commit includes an important change. Duplicated complete_wbs_code is forbidden. To ensure this is respected I made this module dependent on account_analytic_sequence. It makes sense to generate automated sequences when using WBS. To avoid crashing on installation I added a hook that assigns a proper code to every WBS element.

Regards.

**Comment by AaronHForgeFlow** [comment]
The last commit changes accordingly the complete_wbs_code when the parent project changes.

**Comment by AaronHForgeFlow** [comment]
Currently, it is quite nasty the use of this module together with project_key. Adding the same name is a possibility in WBS:

Project1/Distribution
Project2/Distribution

With both modules installed we lose the ability of creating fast projects in the list view as this module proposes.

In order to manage both modules in the same environment a glue module will be needed.

Regards.

File: project_wbs/models/project_project.py
Comment (nikul-serpentcs) [inline project_wbs/models/project_project.py:99]: @aheficent IMO I think used False like `project_item.project_child_complete_ids = child_ids or False`

File: project_wbs/models/project_project.py
Comment (nikul-serpentcs) [inline project_wbs/models/project_project.py:156]: @aheficent IMO I think here create one general method, because this logic use many time.

2> https://github.com/OCA/project/pull/416/files#diff-836e3a974ffa8d76a81bfa20c5111dfaR209
3> https://github.com/OCA/project/pull/416/files#diff-836e3a974ffa8d76a81bfa20c5111dfaR233

File: project_wbs/view/account_analytic_account_view.xml
Comment (nikul-serpentcs) [inline project_wbs/view/account_analytic_account_view.xml]: @aheficent remove ` <data>`

File: project_wbs/view/project_project_view.xml
Comment (nikul-serpentcs) [inline project_wbs/view/project_project_view.xml]: @aheficent IMHO Directly use `<field>` tag.

File: project_wbs/view/project_project_view.xml
Comment (nikul-serpentcs) [inline project_wbs/view/project_project_view.xml]: @aheficent same

File: project_wbs/view/project_project_view.xml
Comment (nikul-serpentcs) [inline project_wbs/view/project_project_view.xml]: @aheficent same

File: project_wbs/view/project_project_view.xml
Comment (nikul-serpentcs) [inline project_wbs/view/project_project_view.xml]: @aheficent same and change all xpath if not use direct field.

File: project_wbs/models/project_project.py
Comment (AaronHForgeFlow) [reply project_wbs/models/project_project.py:99]: if no child_ids then an empty record would be assigned. That would not trigger any error.

File: project_wbs/models/project_project.py
Comment (AaronHForgeFlow) [reply project_wbs/models/project_project.py:156]: I am not sure, because it is just a single search.

File: project_wbs/view/account_analytic_account_view.xml
Comment (AaronHForgeFlow) [reply project_wbs/view/account_analytic_account_view.xml]: ok

**Comment by bennals** [comment]
I have encountered an issue with the complete wbs code recently.  When changing the parent analytical account of a project that has child analytical accounts, the complete wbs codes of those child analytical accounts do not get automatically updated.

**Comment by AaronHForgeFlow** [comment]
HI @bennals  Thank you for reporting this issue. I'll commit the fix soon, in both v11 and v10 at #427

**Comment by AaronHForgeFlow** [comment]
HI @bennals the last commit means to solve that issue.

Regards

**Comment by bennals** [comment]
Thanks @aheficent. I'll test and report back.

**Comment by alexey-pelykh** [comment]
Any help needed here?

**Comment by AaronHForgeFlow** [comment]
@alexey-pelykh I think this is ready to review, we need reviewers.

File: oca_dependencies.txt
Comment (alexey-pelykh) [inline oca_dependencies.txt]: Is it https://github.com/OCA/account-analytic/tree/11.0/account_analytic_parent?

File: oca_dependencies.txt
Comment (alexey-pelykh) [inline oca_dependencies.txt]: Same here https://github.com/OCA/account-analytic/tree/11.0/account_analytic_sequence ?

File: oca_dependencies.txt
Comment (AaronHForgeFlow) [reply oca_dependencies.txt]: Yes it is :) Just dropped the commit that added the dependency

File: oca_dependencies.txt
Comment (AaronHForgeFlow) [reply oca_dependencies.txt]: I just removed the dependency. Thanks.

**Review by alexey-pelykh — APPROVED** [review]

**Comment by alexey-pelykh** [comment]
@aheficent there's a codecov degradation, please either cover code with more tests or use
```
# pragma: no cover
```

**Comment by alexey-pelykh** [comment]
@aheficent eagerly waiting for this to be fixed (and merged afterwards) to port over to 12 👍

**Comment by AaronHForgeFlow** [comment]
@alexey-pelykh Ok! I'll try to have a look today

**Comment by alexey-pelykh** [comment]
@aheficent please check https://codecov.io/gh/OCA/project/compare/b0cb2df9318c32ba64f4512bbc7818c3f3694b07...4f805990284f168da3bcba0643bfc7daae3b953d/diff to see which cases should be added

**Comment by AaronHForgeFlow** [comment]
Finally :100: % haha

**Comment by bennals** [comment]
@aheficent - Found another small bug with the complete_wbs_code updating.

When a project that has a parent analytic account (and therefore a complete_wbs_code that includes the parent's code) has it's parent removed (effectively making it a top level account), the complete_wbs_code does not get updated to reflect that.  Instead, it retains the wbs code from the account that was previously it's parent in it's complete_wbs_code.  Obviously this could lead to incorrect filtering etc.

**Comment by AaronHForgeFlow** [comment]
Hi @bennals I tested and you are right. Thanks for reporting. I'll fix that.

**Comment by AaronHForgeFlow** [comment]
Not sure why runbot is red. Seems unrelated. I'll keep checking.
`account.analytic.account.display_name depends on itself ; please fix its decorator @api.depends().`

Could not reproduce in local

**Comment by ovnicraft** [comment]
@aheficent runbot is red, its related this migration ? or this is ready to merge ?

**Comment by AaronHForgeFlow** [comment]
@ovnicraft I think it is because of this module (something related to the hierarchy), I'm still trying to find what is the problem. The error message it is quite misleading: https://github.com/OCA/project/pull/416#issuecomment-452772707

**Comment by bennals** [comment]
@aheficent I think I found another bug though I can't connect to runbot to check.

Locally, I've found that duplicating a project with tasks after navigating to the project settings form via the drill down buttons in the wbs list view throws the following error.

`2019-05-14 21:43:17,358 25078 ERROR production odoo.sql_db: bad query: b'INSERT INTO "project_task" ("id", "p_loop", "lunch_break_finish7", "workshop_start7", "workshop_start6", "color_gantt", "ptr_parent_id", "parent_id", "workshop_start11", "trip_kms", "displayed_image_id", "remaining_hours", "sorting_level", "date_assign", "color", "plan_duration", "sale_line_id", "workshop_finish9", "site_email", "kanban_state", "no_of_participants", "lunch_break_finish6", "workshop_start17", "fold_child", "description", "planned_hours", "workshop_finish7", "lunch_break_start1", "workshop_finish18", "packing_notes", "no_of_facilitators", "timebox_id", "workshop_start13", "message_last_post", "detail_plan", "themes_requirements", "workshop_start8", "lunch_break_finish2", "workshop_finish5", "sequence", "date_last_stage_update", "date_start", "workshop_finish10", "workshop_finish8", "lunch_break_start4", "notes", "facilitators_notes", "workshop_finish19", "lunch_break_start6", "constrain_type", "critical_path", "workshop_finish11", "template_task", "fixed_calc_type", "on_gantt", "user_id", "lunch_break_start5", "workshop_finish1", "age_group_grade", "sorting_level_seq", "workshop_finish17", "lunch_break_finish1", "site_contact", "workshop_start10", "workshop_start2", "workshop_start3", "workshop_start4", "workshop_start12", "lunch_break_start3", "date_finished", "email_from", "workshop_finish2", "email_cc", "lunch_break_start2", "no_of_workshops", "site_number", "workshop_start15", "workshop_start20", "additional_info", "priority", "workshop_start18", "parking_info", "workshop_finish15", "workshop_start9", "lunch_break_start7", "company_id", "workshop_start14", "stage_id", "workshop_finish13", "schedule_mode", "workshop_finish16", "project_id", "workshop_finish3", "trip_time", "workshop_start5", "workshop_finish21", "workshop_finish4", "lunch_break_finish3", "constrain_date", "is_milestone", "lunch_break_finish5", "workshop_start21", "sorting_seq", "name", "workshop_finish20", "fold_self", "active", "partner_id", "workshop_location", "color_gantt_set", "date_end", "workshop_start16", "final_participants", "trolley_access", "workshop_finish12", "x_role", "context_id", "workshop_start1", "workshop_finish14", "lunch_break_finish4", "info", "workshop_finish6", "workshop_start19", "create_uid", "write_uid", "create_date", "write_date") VALUES(nextval(\'project_task_id_seq\'), false, NULL, NULL, NULL, \'rgba(170,170,13,0.53)\', NULL, 3512, NULL, NULL, NULL, \'0.00\', 0, \'2019-05-14 21:43:17\', 0, 10800, NULL, NULL, NULL, \'normal\', 0, NULL, NULL, false, \'<p><br></p>\', 0.0, NULL, NULL, NULL, NULL, 0, NULL, NULL, NULL, false, NULL, NULL, NULL, NULL, 10, \'2019-05-14 21:43:17\', \'2019-05-13 18:45:00\', NULL, NULL, NULL, NULL, NULL, NULL, NULL, \'snet\', false, NULL, false, \'work\', true, 1, NULL, NULL, NULL, 0, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, 0, NULL, NULL, NULL, NULL, \'0\', NULL, NULL, NULL, NULL, NULL, 1, NULL, 112, NULL, \'auto\', NULL, 5199, NULL, NULL, NULL, NULL, NULL, NULL, \'2019-04-01 14:00:00\', false, NULL, NULL, 1, \'test task for reader\', NULL, false, true, NULL, NULL, false, \'2019-05-13 19:45:00\', NULL, 0, NULL, NULL, NULL, NULL, NULL, NULL, NULL, 0, NULL, NULL, 127, 127, (now() at time zone \'UTC\'), (now() at time zone \'UTC\')) RETURNING id'
ERROR: insert or update on table "project_task" violates foreign key constraint "project_task_parent_id_fkey"
DETAIL:  Key (parent_id)=(3512) is not present in table "project_task".`

Screen Shot 2019-05-15 at 7 44 15 AM

If I refresh the browser window after this error happens the project can be duplicated without any problem.  Also, if I navigate to the project settings form **without** using the drill down buttons (filtering a list or progressively using the project settings buttons in the 'work breakdown structure' tab) then duplicating a project with tasks is fine.

As you can see, I have modules installed that create a lot of custom fields in project_task, but the error is reproducible if I uninstall those modules.

Looking forward to testing on runbot when it's possible and hope to see this reviewed and (together with project_wbs_task) made part of the official OCA project repositry.

**edit** I just noticed there's some formatting code in the middle of that error message after pasting it into the comment here and forgetting to escape it.   This bit - `\'<p><br></p>\.`  Is that meant to be there?

**Comment by AaronHForgeFlow** [comment]
@bennals we will push to get this merged. But needs to fix before the issue you just commented. We will check what is going on. 

After this is fixed you can also approve the PR

Regards.

**Comment by bennals** [comment]
Ok, thanks Aaron.  I might not be the best person to approve a PR though as I can’t code to save my myself.

**Comment by AaronHForgeFlow** [comment]
@bennals after the latest changes in #417 I think this is issue is resolved. Thanks.

@nikul-serpentcs can you please update you review? Thanks

**Review by nikul-serpentcs — APPROVED** [review]
Only Code Review LGTM :+1:

**Comment by rafaelbn** [comment]
Please @aheficent could you rebase and check that all bots are green :green_heart: ? Thanks!

**Comment by MiquelRForgeFlow** [comment]
I think that the problem in runbot is due to `account_analytic_parent`. In that module, the `name_get` method is defined without `super`. And in this `project_wbs` module, the `name_get` also is defined without calling `super`. Perhaps this creates an interference.

**Review by rafaelbn — DISMISSED** [review]
Bots are red :red_circle: please fix it, thanks!

**Comment by bennals** [comment]
Having the wbs element's name in the breadcrumb trail is a big improvement with this latest push @aheficent .  Working well as far as I can tell.  Thank you!

**Comment by AaronHForgeFlow** [comment]
now_what

**Review by NachoAlesLopez — APPROVED** [review]
Code and functional review: OK.

**Comment by AaronHForgeFlow** [comment]
For some reason runbot is not in the checks, but I can tell I would be green :smile:

**Review by MiquelRForgeFlow — APPROVED** [review]

**Comment by AaronHForgeFlow** [comment]
@rafaelbn as long as this is not red anymore, I understand we can move forward with this. Do you agree?

**Comment by AaronHForgeFlow** [comment]
@rafaelbn Is it ok to you to move forward? I understand the only pending reason was CI and it is solved.

**Comment by bennals** [comment]
Well done @AaronHForgeFlow !  Thanks for staying the course:)

**Pull request merged**
